### PR TITLE
[JENKINS-4478] - Move TCP port out from under security

### PR DESCRIPTION
**Pull request by ksenia-nenasheva**
# Description

See [JENKINS-4478](https://issues.jenkins-ci.org/browse/JENKINS-4478).

Details: 
- Moved Agent section out from security checkbox
- Added sections for "Markup Formatter" and "CSRF Protection"

### Changelog entries

Proposed changelog entries:

* Entry 1: Move TCP port out from under security

### CODE REVIEW

**Review by daniel-beck — APPROVED** [review]
This is so great, and many years overdue. Thanks for taking this on!

Has this been tested manually or via ATH to make sure no UI/form weirdness results?

**Comment by ksenia-nenasheva** [comment]
It has been tested manually.

**Comment by oleg-nenashev** [comment]
I would expect the change to break some ATH tests and its configuration engine. But it should not be a big deal afterwards, especially since we have plenty of time before the next LTS baseline

**Review by oleg-nenashev — APPROVED** [review]

**Comment by recena** [comment]
@ksenia-nenasheva Please, for UI changes, provides a screenshot BEFORE / AFTER the changes.

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy [hunk 1]
```diff
@@ -26,6 +26,23 @@ l.layout(norefresh:true, permission:app.ADMINISTER, title:my.displayName, csscla
             set("descriptor", my.descriptor);
 
             f.optionalBlock( field:"useSecurity", title:_("Enable security"), checked:app.useSecurity) {
+                f.entry (title:_("Disable remember me"), field: "disableRememberMe") {
+                    f.checkbox()
+                }
+
+                f.entry(title:_("Access Control")) {
+                    table(style:"width:100%") {
+                        f.descriptorRadioList(title:_("Security Realm"),varName:"realm",         instance:app.securityRealm,         descriptors:SecurityRealm.all())
```
Comment (recena) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: I don't understand this code formatting, why not?

```
f.descriptorRadioList(title: _("Security Realm"), varName: "realm", instance: app.securityRealm, descriptors:SecurityRealm.all())
```

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy [hunk 2]
```diff
@@ -26,6 +26,23 @@ l.layout(norefresh:true, permission:app.ADMINISTER, title:my.displayName, csscla
             set("descriptor", my.descriptor);
 
             f.optionalBlock( field:"useSecurity", title:_("Enable security"), checked:app.useSecurity) {
+                f.entry (title:_("Disable remember me"), field: "disableRememberMe") {
+                    f.checkbox()
+                }
+
+                f.entry(title:_("Access Control")) {
+                    table(style:"width:100%") {
```
Comment (recena) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: Bad idea to use inline CSS style but certainly Jenkins project is full of them 😢

**Comment by recena** [comment]
The UI change LGTM

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy (hunk 2 above)
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: I agree. Tables are also not the best practice. But it is a moved code anyway

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy (hunk 1 above)
Comment (daniel-beck) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: @recena Note that is was this weird before, and @ksenia-nenasheva just didn't change it.

**Comment by jglick** [comment]
Does this fix [JENKINS-40228](https://issues.jenkins-ci.org/browse/JENKINS-40228)?

**Comment by jglick** [comment]
Why not just get rid of the `useSecurity` checkbox altogether and have all the controls at top level? As of 2.0, security is enabled by default—you need to go out of your way to disable it. If you do so, you will just have selected some other insecure options for the security realm and authorization strategy; the checkbox adds no value that I can see.

**Comment by daniel-beck** [comment]
> Why not just get rid of the `useSecurity` checkbox altogether and have all the controls at top level?

I would expect that to not be that simple. For example, people skipping the wizard and getting no security by default (e.g. in Puppet managed environments). Additionally, we may need to give `SecurityRealm.NO_AUTHENTICATION` an `@Extension`, and possibly adapt authn/authz option related methods in `Jenkins` (e.g. `Jenkins#disableSecurity()`).

Therefore I would prefer to log this as an issue, and leave it out of this PR.

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy (hunk 1 above)
Comment (recena) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: @daniel-beck The diff those lines were modified...

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy (hunk 2 above)
Comment (recena) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: @oleg-nenashev Tables are not good when are using in a wrong way 😄

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy (hunk 1 above)
Comment (daniel-beck) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: @recena The lines [67-76](https://github.com/jenkinsci/jenkins/pull/2900/files#diff-459faeca0eeaad7c70ad86b5eb914055L67) were moved to [29-38](https://github.com/jenkinsci/jenkins/pull/2900/files#diff-459faeca0eeaad7c70ad86b5eb914055R29) without modification.

Hence the answer for your question…

> I don't understand this code formatting, why not [this other way]?

… is: "It has been that way before and left unchanged".

Comment (recena) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: @daniel-beck If you move a block of code, and you can apply a better code-style format....why not? the diff will be as complex 😄

Comment (daniel-beck) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]: @recena Not saying it shouldn't have been done (it probably should, but not actively making it worse isn't reason to reject), but you asked "Why not use sane formatting?" and I provided the answer.

**Comment by jglick** [comment]
@daniel-beck is it really that difficult? From a quick attempt,

```diff
diff --git a/core/src/main/java/hudson/security/GlobalSecurityConfiguration.java b/core/src/main/java/hudson/security/GlobalSecurityConfiguration.java
index b9c6fee0de..9fb74c7344 100644
--- a/core/src/main/java/hudson/security/GlobalSecurityConfiguration.java
+++ b/core/src/main/java/hudson/security/GlobalSecurityConfiguration.java
@@ -111,8 +111,7 @@ public class GlobalSecurityConfiguration extends ManagementLink implements Descr
         // for compatibility reasons, the actual value is stored in Jenkins
         Jenkins j = Jenkins.getInstance();
         j.checkPermission(Jenkins.ADMINISTER);
-        if (json.has("useSecurity")) {
-            JSONObject security = json.getJSONObject("useSecurity");
+            JSONObject security = json; // TODO inline
             j.setDisableRememberMe(security.optBoolean("disableRememberMe", false));
             j.setSecurityRealm(SecurityRealm.all().newInstanceFromRadioList(security, "realm"));
             j.setAuthorizationStrategy(AuthorizationStrategy.all().newInstanceFromRadioList(security, "authorization"));
@@ -135,9 +134,6 @@ public class GlobalSecurityConfiguration extends ManagementLink implements Descr
                 }
             }
             j.setAgentProtocols(agentProtocols);
-        } else {
-            j.disableSecurity();
-        }
 
         if (json.has("markupFormatter")) {
             j.setMarkupFormatter(req.bindJSON(MarkupFormatter.class, json.getJSONObject("markupFormatter")));
diff --git a/core/src/main/java/hudson/security/SecurityRealm.java b/core/src/main/java/hudson/security/SecurityRealm.java
index 6cb69906e0..bd6e5fabd8 100644
--- a/core/src/main/java/hudson/security/SecurityRealm.java
+++ b/core/src/main/java/hudson/security/SecurityRealm.java
@@ -69,6 +69,8 @@ import java.io.UnsupportedEncodingException;
 import java.util.List;
 import java.util.Map;
 import java.util.logging.Logger;
+import net.sf.json.JSONObject;
+import org.jenkinsci.Symbol;
 
 /**
  * Pluggable security realm that connects external user database to Hudson.
@@ -550,15 +552,6 @@ public abstract class SecurityRealm extends AbstractDescribableImpl<SecurityReal
         }
 
         /**
-         * This special instance is not configurable explicitly,
-         * so it doesn't have a descriptor.
-         */
-        @Override
-        public Descriptor<SecurityRealm> getDescriptor() {
-            return null;
-        }
-
-        /**
          * There's no group.
          */
         @Override
@@ -580,6 +573,23 @@ public abstract class SecurityRealm extends AbstractDescribableImpl<SecurityReal
         private Object readResolve() {
             return NO_AUTHENTICATION;
         }
+
+        @Extension
+        @Symbol("none")
+        public static class DescriptorImpl extends Descriptor<SecurityRealm> {
+
+            @Override
+            public String getDisplayName() {
+                return "None"; // TODO I18N
+            }
+
+            @Override
+            public SecurityRealm newInstance(StaplerRequest req, JSONObject formData) throws FormException {
+                return NO_AUTHENTICATION;
+            }
+
+        }
+
     }
 
     /**
diff --git a/core/src/main/java/jenkins/model/Jenkins.java b/core/src/main/java/jenkins/model/Jenkins.java
index f187137018..57fc1afec6 100644
--- a/core/src/main/java/jenkins/model/Jenkins.java
+++ b/core/src/main/java/jenkins/model/Jenkins.java
@@ -2585,6 +2585,7 @@ public class Jenkins extends AbstractCIBase implements DirectlyModifiableTopLeve
         this.disableRememberMe = disableRememberMe;
     }
 
+    @Deprecated
     public void disableSecurity() {
         useSecurity = null;
         setSecurityRealm(SecurityRealm.NO_AUTHENTICATION);
diff --git a/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy b/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy
index f082a3fce5..913d208583 100644
--- a/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy
+++ b/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy
@@ -25,7 +25,6 @@ l.layout(norefresh:true, permission:app.ADMINISTER, title:my.displayName, csscla
             set("instance",my);
             set("descriptor", my.descriptor);
 
-            f.optionalBlock( field:"useSecurity", title:_("Enable security"), checked:app.useSecurity) {
                 f.entry(title: _("TCP port for JNLP agents"), field: "slaveAgentPort") {
                     if (my.slaveAgentPortEnforced) {
                         if (my.slaveAgentPort == -1) {
@@ -74,7 +73,6 @@ l.layout(norefresh:true, permission:app.ADMINISTER, title:my.displayName, csscla
                         f.descriptorRadioList(title:_("Authorization"), varName:"authorization", instance:app.authorizationStrategy, descriptors:AuthorizationStrategy.all())
                     }
                 }
-            }
 
             f.dropdownDescriptorSelector(title:_("Markup Formatter"),descriptors: MarkupFormatterDescriptor.all(), field: 'markupFormatter')
 
```

seems to work.

**Comment by ksenia-nenasheva** [comment]
**Before:**

- Disable security:



- Enable security:


**After:**

**Comment by ksenia-nenasheva** [comment]
Hi @daniel-beck, @jglick 
Yes, I could work to it. But I am not sure that it should be in this PR. I prefer different PRs for different bugs.
Could you please make a decision?

**Comment by daniel-beck** [comment]
@jglick @ksenia-nenasheva I would also prefer if the scope of this PR not be extended. This has merit with or without a followup change due to new section headers for JNLP port and CSRF protection.

**Review by oleg-nenashev — APPROVED** [review]
I agree with the merge in the current state. I am not 100% sure it qualifies for backporting (if somebody submits the manually-generated form in his configuration-as-code engine), but the change addresses what is required in JENKINS-4478.

After the merge we should reopen https://issues.jenkins-ci.org/browse/JENKINS-40228, which is not a full duplicate.

**Pull request merged**